docs/proposals/kubelet-tls-bootstrap.md

```diff
@@ -0,0 +1,231 @@
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubelet TLS bootstrap
+
+Author: George Tankersley (george.tankersley@coreos.com)
+
+## Preface
+
+This document describes a method for a kubelet to bootstrap itself
+into a TLS-secured cluster. Crucially, it automates the provision and
+distribution of signed certificates.
+
+## Overview
+
+When a kubelet runs for the first time, it must be given TLS assets
+or generate them itself. In the first case, this is a burden on the cluster
+admin and a significant logistical barrier to secure Kubernetes rollouts. In
+the second, the kubelet must self-sign its certificate and forfeits many of the
+advantages of a PKI system. Instead, we propose that the kubelet generate a
+private key and a CSR for submission to a cluster-level certificate signing
+process.
+
+## Preliminaries
+
+We assume the existence of a functioning control plane. The
+apiserver should be configured for TLS initially or possess the ability to
+generate valid TLS credentials for itself. If secret information is passed in
+the request (e.g. auth tokens supplied with the request or included in
+ExtraInfo) then all communications from the node to the apiserver must take
+place over a verified TLS connection.
+
+Each node is additionally provisioned with the following information:
+
+1. Location of the apiserver
+2. Any CA certificates necessary to trust the apiserver's TLS certificate
+3. Access tokens (if needed) to communicate with the CSR endpoint
+
+These should not change often and are thus simple to include in a static
+provisioning script.
+
+## API Changes
+
+### CertificateSigningRequest Object
+
+We introduce a new API object to represent PKCS#10 certificate signing
+requests. It will be accessible under:
+
+`/api/vX/certificaterequests/mycsr`
+
+It will have the following structure:
+
+```go
+// Describes a certificate signing request
+type CertificateSigningRequest struct {
+        api.TypeMeta         `json:",inline"`
+        api.ObjectMeta       `json:"metadata,omitempty"`
+
+        // Specifies the behavior of the CSR
+        Spec CertificateSigningRequestSpec
+
+        // Most recently observed status of the CSR
+        Status CertificateSigningRequestStatus
+}
+
+type CertificateSigningRequestSpec struct {
+        // Raw PKCS#10 CSR data
+        CertificateRequest []byte
+
+        // Fingerprint of the public key that signed the CSR
+        Fingerprint string
+
+        // Subject fields from the CSR
+        Subject pkix.Name
+
+        // DNS SANs from the CSR
+        Hostnames []string
+
+        // IP SANs from the CSR
+        IPAddresses []string
+
+        // Extra information the node wishes to send with the request
+        ExtraInfo []string
+}
+
+type CertificateSigningRequestStatus struct {
```

Comment:
My goal is to avoid having to provision a node with anything other than the location of its apiserver (or even a bootstrap server as in https://github.com/kubernetes/kubernetes/issues/23172). Given that, a node will be able to request a certificate with this API, then switch to TLS auth with meaningful subject information for further communication (including registration) with the apiserver. This allows you to write more robust ABAC policy and enforce clustering restrictions with arbitrary admission controllers. We're building one based on TPM attestation, but it could be anything.

In this scenario, I'm not sure where a username would come from. A field for any relevant caller ID sounds very useful though- for direct API access (kubectl/curl) I'd expect client credentials, for services in pods I'd expect a service account, etc. Would that do what you need?